tap_harvest/streams/abstracts.py

```diff
@@ -0,0 +1,281 @@
+            if record[key] is not None:
+                record[key + "_id"] = record[key]["id"]
+            else:
+                record[key + "_id"] = None
+
+        return record
+
+    def modify_object(self, record: Dict, parent_record: Dict = None) -> Dict:
+        """Modify the record before writing to the stream."""
+        record = self.add_object_to_id(record)
+        return record
+
+    def append_times_to_dates(self, record: Dict):
+        """Append times to date fields."""
+        for date_field in self.date_fields:
+            if record.get(date_field):
+                record[date_field] = strftime(strptime_with_tz(record[date_field]))
+
+    def get_url_endpoint(self, parent_obj: Dict = None) -> str:
+        """Get the URL endpoint for the stream."""
+        return self.url_endpoint or f"{self.client.base_url}/{self.path}"
+
+
+class IncrementalStream(BaseStream):
+    """Base Class for Incremental Stream."""
+
+    def get_bookmark(self, state: dict, stream: str, key: Any = None) -> int:
+        """A wrapper for singer.get_bookmark to deal with compatibility for
+        bookmark values or start values."""
+        return get_bookmark(
+            state,
+            stream,
+            key or self.replication_keys[0],
+            self.client.config["start_date"],
+        )
+
+    def write_bookmark(
+        self, state: dict, stream: str, key: Any = None, value: Any = None
+    ) -> Dict:
+        """A wrapper for singer.get_bookmark to deal with compatibility for
+        bookmark values or start values."""
+        if not (key or self.replication_keys):
+            return state
+
+        current_bookmark = get_bookmark(
+            state,
+            stream,
+            key or self.replication_keys[0],
+            self.client.config["start_date"],
+        )
+        value = max(current_bookmark, value)
+        return write_bookmark(state, stream, key or self.replication_keys[0], value)
+
+    def sync(
+        self,
+        state: Dict,
+        transformer: Transformer,
+        parent_obj: Dict = None,
+    ) -> Dict:
+        """Implementation for `type: Incremental` stream."""
+        bookmark_date = self.get_bookmark(state, self.tap_stream_id)
+        current_max_bookmark_date = bookmark_date
+
+        self.update_params(updated_since=bookmark_date)
+        self.url_endpoint = self.get_url_endpoint(parent_obj)
+
+        with metrics.record_counter(self.tap_stream_id) as counter:
+            for record in self.get_records():
+                record = self.modify_object(record, parent_obj)
+                transformed_record = transformer.transform(
+                    copy.deepcopy(record), self.schema, self.metadata
+                )
+                self.append_times_to_dates(transformed_record)
+
+                record_timestamp = transformed_record[self.replication_keys[0]]
+                if record_timestamp >= bookmark_date:
+                    if self.is_selected():
```

Comment:
We should add appropriate why we have this condition in refernce to parent-child relationship.